packages/apollo-federation/src/service/buildFederatedSchema.ts

```diff
@@ -1,100 +1,45 @@
-import {
-  DocumentNode,
-  GraphQLSchema,
-  isObjectType,
-  isUnionType,
-  GraphQLUnionType,
-  GraphQLObjectType,
-  specifiedDirectives,
-} from 'graphql';
+import { DocumentNode, GraphQLSchema, specifiedDirectives } from 'graphql';
 import {
   buildSchemaFromSDL,
-  transformSchema,
   GraphQLSchemaModule,
   modulesFromSDL,
-  addResolversToSchema,
+  GraphQLResolverMap,
 } from 'apollo-graphql';
-import federationDirectives, { typeIncludesDirective } from '../directives';
-
-import { serviceField, entitiesField, EntityType } from '../types';
-
-import { printSchema } from './printFederatedSchema';
+import federationDirectives from '../directives';
 
 import 'apollo-server-env';
+import { transformFederatedSchema } from './transformFederatedSchema';
+import { extractFederationResolvers } from './extractFederationResolvers';
 
 export function buildFederatedSchema(
-  modulesOrSDL: (GraphQLSchemaModule | DocumentNode)[] | DocumentNode,
+  modulesOrSDLOrSchema:
+    | (GraphQLSchemaModule | DocumentNode)[]
+    | DocumentNode
+    | GraphQLSchema,
 ): GraphQLSchema {
-  const modules = modulesFromSDL(modulesOrSDL);
+  // Extract federation specific resolvers from already constructed
+  // GraphQLSchema and transform it to a federated schema.
+  if (modulesOrSDLOrSchema instanceof GraphQLSchema) {
+    return transformFederatedSchema(modulesOrSDLOrSchema, [
+      extractFederationResolvers(modulesOrSDLOrSchema),
+    ]);
+  }
 
-  let schema = buildSchemaFromSDL(
-    modules,
-    new GraphQLSchema({
-      query: undefined,
-      directives: [...specifiedDirectives, ...federationDirectives],
-    }),
-  );
+  // Transform *modules* or *sdl* into a federated schema.
+  const modules = modulesFromSDL(modulesOrSDLOrSchema);
 
-  // At this point in time, we have a schema to be printed into SDL which is
-  // representative of what the user defined for their schema. This is before
-  // we process any of the federation directives and add custom federation types
-  // so its the right place to create our service definition sdl.
-  //
-  // We have to use a modified printSchema from graphql-js which includes
-  // support for preserving the *uses* of federation directives while removing
-  // their *definitions* from the sdl.
-  const sdl = printSchema(schema);
+  const resolvers = modules
```

Comment:
Alternatively, we could merge resolvers from all modules here, so we can simplify `transformFederatedSchema`.